pkg/ssh/metrics.go

```diff
@@ -16,5 +28,81 @@ func newPromMetrics() *promMetrics {
 			},
 			[]string{"exit_code"},
 		),
+		channelsCount: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name:      "ssh_channels",
+				Help:      "Number of active SSH channels",
+				Namespace: "pdc_agent",
+			},
+		),
+		tcpConnectionsCount: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name:      "tcp_connections",
+				Help:      "Number of open TCP connections",
+				Namespace: "pdc_agent",
+			},
+			[]string{"target", "status"},
+		),
+		timeToConnect: metrics.NewNativeHistogramVec(
+			prometheus.HistogramOpts{
+				Name:      "ssh_time_to_connect_seconds",
+				Help:      "Time spent to establish SSH connection",
+				Namespace: "pdc_agent",
+			},
+			[]string{"event"},
+		),
+	}
+}
+
+type logMetricsParser struct {
+	m           *promMetrics
+	connStart   time.Time
+	connRestart time.Time
+	isStart     bool // to differentiate between "start" and "restart" connection
+}
+
+func (p logMetricsParser) parseLogMetrics(msg []byte) {
+	switch {
+	case bytes.Contains(msg, []byte("nchannels")):
+		p.channelsCount(msg)
+	case bytes.Contains(msg, []byte("connected to")):
+		pattern := `connected to (.+?) port (\d+)`
+		p.tcpConnCount(msg, pattern, "success")
+	case bytes.Contains(msg, []byte("connect_to")):
+		pattern := `connect_to (.+?) port (\d+): failed.`
+		p.tcpConnCount(msg, pattern, "failure")
+	case bytes.Contains(msg, []byte("This is Grafana Private Datasource Connect!")):
+		if p.isStart {
+			p.m.timeToConnect.
+				WithLabelValues("start").
+				Observe(time.Since(p.connStart).Seconds())
+		} else {
+			p.m.timeToConnect.
+				WithLabelValues("restart").
+				Observe(time.Since(p.connRestart).Seconds())
+		}
+	}
+}
+
+func (p logMetricsParser) channelsCount(msg []byte) {
+	re := regexp.MustCompile(`nchannels (\d+)`)
```

Comment:
I think we might be able to make some performance savings by compiling these regexes once, and then using the compiled regex multiple time.

Reply:
you could check this with a benchmark test if you like.